### PR TITLE
SEQNG-1067 Tcs in-position stabilization time now depends on magnitude of movement.

### PR DESCRIPTION
**Pull request by jluhrs**
The TCS in-position signal takes some time to be stable. This PR makes the time to wait for it to stabilize depend on the subsystems used and the magnitude of the displacement.
The parameters for the calculations are placeholders until they are measured. I don't intend to merge the PR until then.

### CODE REVIEW

File: modules/seqexec/server/src/main/scala/seqexec/server/tcs/TcsSettleTimeCalculator.scala [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+object TcsSettleTimeCalculator {
+
+  trait SettleTimeCalculator {
+    def calc(displacement: Angle): Time
+  }
+
+  def constantSettleTime(cnst: Time): SettleTimeCalculator = (_: Angle) => cnst
+
+  // Settle time proportional to displacement
+  def linearSettleTime(scale: SettleTimeScale): SettleTimeCalculator = (displacement: Angle) => scale * displacement
```
Comment (jluhrs) [inline modules/seqexec/server/src/main/scala/seqexec/server/tcs/TcsSettleTimeCalculator.scala]: I may need to implement more formulae, like quadratic or exponential, depending on the results of measuring the stabilization times.

File: modules/seqexec/server/src/main/scala/seqexec/server/tcs/TcsNorthControllerEpicsAo.scala [hunk 2]
```diff
@@ -94,14 +95,17 @@ object TcsNorthControllerEpicsAo {
 
     def sysConfig(current: EpicsTcsAoConfig): IO[EpicsTcsAoConfig] = {
       val params = configParams(current)
+      val stabilizationTime = tcs.tc.offsetA
```
Comment (cquiroz) [inline modules/seqexec/server/src/main/scala/seqexec/server/tcs/TcsNorthControllerEpicsAo.scala]: This seems duplicated across north/south, could it be factored out?

File: modules/acm/src/main/java/edu/gemini/epics/acm/CaWindowStabilizer.java [hunk 3]
```diff
@@ -147,16 +147,20 @@ public void removeListener(CaAttributeListener<T> listener) {
         notifier.removeListener(listener);
     }
 
-    public CaWindowStabilizer<T> reset() {
+    public CaWindowStabilizer<T> restart(Duration settleTime) {
```
Comment (cquiroz) [inline modules/acm/src/main/java/edu/gemini/epics/acm/CaWindowStabilizer.java]: Maybe we should document this a bit. What does the `settleTime` do?

Comment (jluhrs) [reply modules/acm/src/main/java/edu/gemini/epics/acm/CaWindowStabilizer.java]: Added documentation. `settleTime` is the time a the value of the filtered `CaAttribute` must be unchanged to be set as the output of the filter.

File: modules/seqexec/server/src/main/scala/seqexec/server/tcs/TcsNorthControllerEpicsAo.scala (hunk 2 above)
Comment (jluhrs) [reply modules/seqexec/server/src/main/scala/seqexec/server/tcs/TcsNorthControllerEpicsAo.scala]: Done.

**Comment by jluhrs** [comment]
I put fixed values for the TCS in-position stabilization filter. The values were taken from the fudge factors in the old Seqexec.

File: modules/seqexec/server/src/main/scala/seqexec/server/tcs/TcsEpics.scala [hunk 4]
```diff
@@ -380,29 +384,38 @@ final class TcsEpics[F[_]: Async](epicsService: CaService, tops: Map[String, Str
...
-    .flatMap(waitForValueF(_, "TRUE", timeout,"TCS inposition flag"))
+  // Tcs fudge1 (time to wait for in-position to change to false)
+  private val tcsSettleTime = 2.8.seconds
+
+  // This functions returns a F that, when run, first waits tcsSettleTime to absorb in-position transients, then waits
+  // for the in-position to change to true and stay true for stabilizationTime. It will wait up to `timeout`
+  // seconds for that to happen.
+  def waitInPosition(stabilizationTime: Time, timeout: Time)(implicit T: Timer[F]): F[Unit] =
+    T.sleep(FiniteDuration(tcsSettleTime.toMillis, TimeUnit.MILLISECONDS)) *>
+      Sync[F].delay(filteredInPositionAttr.restart(java.time.Duration.ofMillis(stabilizationTime.toMillis)))
```
Comment (cquiroz) [inline modules/seqexec/server/src/main/scala/seqexec/server/tcs/TcsEpics.scala]: Will `restart` call into `epics`?

File: modules/seqexec/server/src/main/scala/seqexec/server/tcs/TcsEpics.scala [hunk 5]
```diff
@@ -385,12 +389,16 @@ final class TcsEpics[F[_]: Async](epicsService: CaService, tops: Map[String, Str
   private val filteredInPositionAttr: CaWindowStabilizer[String] =
     new CaWindowStabilizer[String](inPositionAttr, java.time.Duration.ofMillis(defaultTcsStabilizeTime.toMillis))
 
+  // Tcs fudge1 (time to wait for in-position to change to false)
+  private val tcsSettleTime = 2.8.seconds
```
Comment (cquiroz) [inline modules/seqexec/server/src/main/scala/seqexec/server/tcs/TcsEpics.scala]: This is from tcl-seqexec right?

File: modules/seqexec/server/src/main/scala/seqexec/server/tcs/TcsEpics.scala (hunk 4 above)
Comment (jluhrs) [reply modules/seqexec/server/src/main/scala/seqexec/server/tcs/TcsEpics.scala]: No, restart resets the filter

File: modules/seqexec/server/src/main/scala/seqexec/server/tcs/TcsEpics.scala (hunk 5 above)
Comment (jluhrs) [reply modules/seqexec/server/src/main/scala/seqexec/server/tcs/TcsEpics.scala]: Yes

**Pull request merged**